src/controllers/hid/hidiothread.h

```diff
@@ -0,0 +1,96 @@
+#pragma once
+
+#include <QAtomicInteger>
+#include <QSemaphore>
+#include <QThread>
+#include <map>
+
+#include "controllers/controller.h"
+#include "controllers/hid/hiddevice.h"
+#include "controllers/hid/hidiooutputreport.h"
+#include "util/compatibility/qmutex.h"
+#include "util/duration.h"
+
+enum class HidIoThreadState {
+    Initialized,
+    OutputActive,
+    InputOutputActive,
+    StopWhenAllReportsSent,
+    StopRequested,
+    Stopped,
+};
+
+class HidIoThread : public QThread {
+    Q_OBJECT
+  public:
+    HidIoThread(hid_device* pDevice,
+            const mixxx::hid::DeviceInfo& deviceInfo);
+    ~HidIoThread() override;
+
+    void run() override;
+
+    /// Sets the state of the HidIoThread lifecycle,
+    /// if the previous state was not the expected,
+    /// it returns with false as result.
+    [[nodiscard]] bool testAndSetThreadState(HidIoThreadState expectedState,
+            HidIoThreadState newState);
+
+    /// Set the state of the HidIoThread lifecycles unconditional
+    void setThreadState(HidIoThreadState expectedState);
+
+    /// Wait's until the run loop stopped, or the specified timeout, is reached.
+    /// Returns immediately with true if the run loop is stopped.
+    [[nodiscard]] bool waitUntilRunLoopIsStopped(unsigned int timeoutMillis);
+
+    void updateCachedOutputReportData(const QByteArray& reportData, unsigned int reportID);
+    QByteArray getInputReport(unsigned int reportID);
+    void sendFeatureReport(const QByteArray& reportData, unsigned int reportID);
+    QByteArray getFeatureReport(unsigned int reportID);
+
+  signals:
+    /// Signals that a HID InputReport received by Interrupt triggered from HID device
+    void receive(const QByteArray& data, mixxx::Duration timestamp);
+
+  private:
+    bool sendNextCachedOutputReport();
+
+    void pollBufferedInputReports();
+    void processInputReport(int bytesRead);
+
+    const mixxx::hid::DeviceInfo m_deviceInfo;
+    const RuntimeLoggingCategory m_logBase;
+    const RuntimeLoggingCategory m_logInput;
+    const RuntimeLoggingCategory m_logOutput;
+
+    /// Mutex must be locked when using the m_pPollData, m_lastPollSize, m_pollingBufferIndex
+    /// or the m_pHidDevice structure, which is not thread-safe for all hidapi backends.
+    QMutex m_hidDeviceMutex;
```

Comment:
The name of the mutex is misleading. The comment explains that it is needed to synchronize access to the whole internal polling state, not only the HID device API.

Ideally one would extract it into a standalone struct or class. Let's simply name it `m_pollHidDeviceMutex`.